processing/src/main/java/org/apache/druid/segment/join/table/BroadcastSegmentIndexedTable.java

```diff
@@ -241,6 +244,22 @@ public ColumnSelectorFactory makeColumnSelectorFactory(ReadableOffset offset, bo
     );
   }
 
+  @Override
+  public Optional<byte[]> computeCacheKey()
+  {
+    SegmentId segmentId = segment.getId();
```

Comment:
are you suggesting to move this logic to `SegmentId` class?

Reply:
I wanted to use `CacheUtil` here but that is in `server` module and not accessible here. May be I can move that class to `core`?

Reply:
`CacheKeyBuilder` requires a prefix byte key in its constructor and hence I was not using it everywhere. I could add some fixed byte key though not sure how much does that help.

Reply:
> `CacheKeyBuilder` requires a prefix byte key in its constructor and hence I was not using it everywhere. I could add some fixed byte key though not sure how much does that help.

This makes me wonder if broadcast segment needs a prefix key. I'm not sure if it would be useful yet. I still find using `CacheKeyBuilder` better as it's less error-prone than computing cache key manually, but I also agree with your point. I will leave this up to you.

Reply:
I think there actually should be a prefix byte here, because the prefix bytes' purpose is to prevent cache key collisions for two implementations that are different but compute cache keys the same way. So each implementation should have its own prefix byte.

I don't have a strong opinion on where the code should live, but it would be good to share it somehow, if that's not too much trouble.

Reply:
Thinking out loud here. For the cache key of queries, I understand that a prefix can help when e.g. a `groupBy` and `TimeSeries` have the same filters, aggregators, etc. For `IndexedTable`, a similar scenario can occur if we have two different implementations of `IndexedTable` that can be invoked on the same segment. when can that occur? May be if the cache is external and the historical/broker is upgraded to a new implementation of `IndexedTable` applicable to that segment? 
I don't mind adding the key prefix. Just trying to understand it better.

Reply:
I'm not sure how there could be cache key collisions between `IndexedTable`s. Even though there could be multiple `IndexedTable` implementations, the way it is used in the callers (`IndexedTableJoinMatcher`, `IndexedTableColumnValueSelector`, etc) should be same across those implementations which will result in the same query result. There will also not be cache key collisions between a broadcast segment (`IndexedTable`) and a regular segment as their cache key computation is different (the datasource name is included in only the former). But, the current way doesn't look very well-structured.

Another approach I've been thinking is making `ReferenceCountedObject` (or `SegmentReference`) cacheable as they represent a segment (either a physical or a virtual). In this way, the cache key for segments will be computed in each `ReferenceCountedObject` implementation and so even `CachingQueryRunner` doesn't have to know how to compute the cache key for segments. The interface to compute a cache key should accept `SegmentDescriptor`. In this case, the prefix will indicate how the segment will be processed (in a hash join or an aggregation without join for now). This approach seems most reasonable to me, but I didn't mention it before since you already went a different way.

Reply:
`ReferenceCountedObject` has wide variety of implementations. I am not sure how can that be made `cacheable` while many of these implementations have nothing to do with query caching. 

how is making `SegmentReference` cacheable different from the very first alternative where we could add `computeCacheKey` method to `Segment`  class?  

> In this case, the prefix will indicate how the segment will be processed (in a hash join or an aggregation without join for now)

By the way, prefix for a join operation is already being used. where do you see this prefix being passed from?

Reply:
> For IndexedTable, a similar scenario can occur if we have two different implementations of IndexedTable that can be invoked on the same segment. when can that occur? May be if the cache is external and the historical/broker is upgraded to a new implementation of IndexedTable applicable to that segment?

I didn’t have a specific scenario in mind, I was just thinking that it’s generally a good practice to have a namespace prefix like this when defining a cache key for an interface implementation. Even if there aren’t others today, it’s possible there will be in the future, and it’s easy for new implementers to forget to update the cache keys of the existing implementations. Better to do it ahead of time.

Reply:
I have added the prefix here